src/Database/Migration.php

```diff
@@ -125,35 +137,39 @@ public static function addSettings(array $defaults)
      */
     public static function addPermissions(array $permissions)
     {
-        $keys = [];
+        $rows = [];
 
         foreach ($permissions as $permission => $groups) {
             foreach ((array) $groups as $group) {
-                $keys[] = [
+                $rows[] = [
                     'group_id' => $group,
                     'permission' => $permission,
                 ];
             }
         }
 
         return [
-            'up' => function (Builder $schema) use ($keys) {
+            'up' => function (Builder $schema) use ($rows) {
                 $db = $schema->getConnection();
 
-                foreach ($keys as $key) {
-                    $instance = $db->table('permissions')->where($key)->first();
+                foreach ($rows as $row) {
+                    if ($db->table('group_permission')->where($row)->exists()) {
+                        continue;
+                    }
 
-                    if (is_null($instance)) {
-                        $db->table('permissions')->insert($key);
+                    if ($db->table('groups')->where('id', $row['group_id'])->doesntExist()) {
+                        continue;
                     }
+
+                    $db->table('group_permission')->insert($row);
                 }
             },
 
-            'down' => function (Builder $schema) use ($keys) {
+            'down' => function (Builder $schema) use ($rows) {
                 $db = $schema->getConnection();
 
-                foreach ($keys as $key) {
-                    $db->table('permissions')->where($key)->delete();
+                foreach ($rows as $row) {
+                    $db->table('group_permission')->where($row)->delete();
```

Comment:
Hmm, this is interesting. Doesn't changing the implementation of this migration helper break old migrations using this type? After all, at the point when those migrations were created (and where they should be run in sequence) the table was still called `permissions`.

That tells me we should inline this kind of migration instead of providing helpers for it.

Reply:
Yes, you're right.

Reply:
@franzliedke This is kind of a pain in the butt to inline. It's quite commonly used, and there's lots of boilerplate required (only add the permission if the group exists, and if it's not already present in the group_permission table).

Could we potentially keep the helper but add a second parameter to indicate the name of the table to use (default to `group_permission`, and old migrations can be updated to pass `permissions`)?

Come to think of it, there is a flaw in our current migrator in that core migrations are run first, then migrations from each extension sequentially... but really they should all be run at the same time in chronological order to prevent issues with this kind of stuff?

Reply:
Okay, but how would we keep that helper for extensions (makes sense), but prevent it being used in core (where table names in migrations should never need to be changed)?

We probably need to add a facility for extensions to reference tables by model, so that their migrations don't break when core changes table names?

By the way, both `groups` and 'group_permission` are referenced in that file.

Reply:
> Okay, but how would we keep that helper for extensions (makes sense), but prevent it being used in core (where table names in migrations should never need to be changed)?

We should probably just inline the helper for the old core migrations.

> We probably need to add a facility for extensions to reference tables by model, so that their migrations don't break when core changes table names?

We could just put the onus on extensions to update their migrations to remain compatible, but I suppose it's pretty easy to do the following:

```php
return Migration::addColumns(User::getTable(), []);
```

Reply:
Hmm, let's avoid referencing model classes in migrations. Extensions should just update the table / column names in their existing migrations - they won't be run again anyway since the timestamps are saved in the database. In the future, things like this won't happen - or if they do, only in major updates, where SemVer should protect the extensions from breaking, as long as they have proper version constraints.

Other than that, :+1: for your poposal.

Reply:
Hmm, we could at least make the logic reusable - this is what I came up with:

~~~php
    public static function addRows($table, array $row, ...$foreignKeys)
    {
        $references = [];

        if (! empty($foreignKeys)) {
            if (count($foreignKeys) % 2 == 1) {
                throw new \InvalidArgumentException(
                    'Two arguments needed for every foreign key'
                );
            }

            while (count($foreignKeys) > 0) {
                $table = array_shift($foreignKeys);
                $keyMap = array_shift($foreignKeys);

                $references[$table] = $keyMap;
            }
        }

        return [
            'up' => function (Builder $schema) use ($table, $row, $references) {
                $db = $schema->getConnection();

                if ($db->table($table)->where($row)->exists()) {
                    return;
                }

                foreach ($references as $table => $keyMap) {
                    $column = key($keyMap);
                    $referencedColumn = current($keyMap);

                    if ($db->table($table)->where($column, $row[$referencedColumn])->doesntExist()) {
                        return;
                    }
                }

                $db->table($table)->insert($row);
            },

            'down' => function (Builder $schema) use ($table, $row) {
                $schema->getConnection()->table($table)->where($row)->delete();
            }
        ];
    }
~~~

You would have to use it like this:
~~~php
Migration::addRow(
    'group_permission',
    [
        'group_id' => Group::MODERATOR_ID,
        'permission' => 'discussion.hidePosts',
    ],
    'groups',
    ['id' => 'group_id']
);
~~~
...with the third and fourth parameter being optional. Not sure if that API is too confusing, or whether we can come up with something better?

Reply:
I've just inlined the core migration in https://github.com/flarum/core/pull/1344/commits/622bf265103e53fe0388b6c97e568010636cbcaa. I think it's simple (and specific) enough that it doesn't warrant a new helper.